### DISCUSSION TITLE
expensive readonly HashSet<T> construction

### DISCUSSION THREAD

**Discussion by kasperk81** [dotnet/runtime#109330] [Q&A]
https://godbolt.org/z/eE5svxTMo

`class A` benefits from JIT optimizations: the array setup is simplified to a direct allocation and pointer assignments, making it efficient with minimal overhead.

In `class B`, using `HashSet<string>` is more costly:
- The JIT initializes hashing and comparison structures and performs helper calls (e.g., `AddIfNotPresent`).
- Each call involves hash calculations and duplicate checks, which the JIT cannot eliminate, leading to higher overhead.

@EgorBo Could the JIT optimize `HashSet` initialization for known unique values, reducing redundant checks?

**Comment by huoyaoyuan** [comment]
> `class A` benefits from JIT optimizations
> Could the JIT optimize `HashSet` initialization for known unique values, reducing redundant checks?

No, because it's not done by the JIT and out of scope.

The work for array is done by C# compiler: https://sharplab.io/#v2:D4AQDABCCMAsDcBYAUCkAmCBBFBvFEhUAzFNGANoC6EAshALwB8BRFrRRARAIYBGAYy4AaDpwhcAHgE8AXiLGcu0dMS5iqSZAF8UaTACE8YkKQASPAM4ALAMoBTAC4AeGGCZ1GLZJ3Y/xhLyCCv4BUnIhAYEqahpaushAA==
JIT is just compiling the IL as-is.

There is optimal building support for `FrozenSet<T>` (#102223), but not `HashSet<T>`.
